Updated the entity names for all z-wave devices.

stefan-jonasson opened the pull request:
Z-Wave have a unique node id, this is now added to the entity name.

This is a change will break old configs.  The Z-Wave entity_id will have to be updated in configurations.

The reason for this change is that the old behavior created non unique entity_ids.
The order devices is included is not always the same so when the service was restarted the entity ids could change and the configuration for entities will be invalid.

There was also inconsistencies between the naming for sensors and switches.

The new entity id naming is as follows: 
`domain.<manufacturer>_<product_name>_<sensor/light/switch-label>_<node_id>`
ex:
`philio_technology_corporation_psm021_slim_multisensor_temperature_7`

If there is several "things" (sensors/switches) in the same node then the 2:nd and up will have their instance id added to the entity_id:
`domain.<manufacturer>_<product_name>_<sensor/light/switch-label>_<node_id>_<instance>`

Ex. For a second temperature sensor in the same device:
`philio_technology_corporation_psm021_slim_multisensor_temperature_7_2`

To update existing configs, the common change will be to append `_<node_id>` to the end of the entity_id refering to all z-wave devices in the config. 

For switch entries, the manufacturer name will also have to be added because the old naming differed against the other domains.

Commit:
Updated the entity names for all z-wave devices.
Z-Wave have a unique node id, this is now added to the entity name.

This is a change will break old configs.  The Z-Wave entity_id will have to be updated in configurations.

The reason for this change is that the old behavior created duplicate entity_ids if you have several similar devices. The order devices is included is not consistent so when the service was restarted the entity ids could change and the configuration for entities will be invalid.

There was also inconsistencies between the naming for sensors and switches.
(cherry picked from commit 33a78df)

rmkraus commented:
Wouldn't it be better to just add the Node ID to the Entity ID rather than to the Name? If the Node ID is added to the name, then wont Z-Wave users be forced to rename every single device in their config?

balloob commented:
I agree with @rmkraus . You can manually set the `entity_id` attribute on an entity to prevent it from being auto generated. The entity_id for Z-Wave devices could be a combination of `node_id` and `command_class`.

stefan-jonasson commented:
Sure I could implement the entity_id property, but wouldn't that also require all z-wave users to update all the z-wave entity_id the config files? 

The reason I changed the name was so I would be a tighter connection between the displayed name and the entity_id. If i were to generate the entity_id from just `node_id` and `command_class` it will just be two numbers and it will be very confusing config files. 
I think it's good that there is a connection between the name you see in the interface and the config values. 
I could generate a similar entity_id as the name so that the names will be as before and the entity_id as like the name but with the appended node_id, what do you think?

balloob commented:
That sounds good. Are names unique per command class? Or does command class
has to be in it?

On Mon, Jan 25, 2016, 02:46 Stefan Jonasson <notifications@github.com>
wrote:

> Sure I could implement the entity_id property, but wouldn't that also
> require all z-wave users to update all the z-wave entity_id the config
> files?
>
> The reason I changed the name was so I would be a tighter connection
> between the displayed name and the entity_id. If i were to generate the
> entity_id from just node_id and command_class it will just be two numbers
> and it will be very confusing config files.
> I think it's good that there is a connection between the name you see in
> the interface and the config values.
> I could generate a similar entity_id as the name so that the names will be
> as before and the entity_id as like the name but with the appended node_id,
> what do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/982#issuecomment-174467396>
> .
>

Commit:
Updated teh entity_id generation

Entity ids will now look like this:
sensor.philio_technology_corporation_psm021_slim_multisensor_doorwindow_sensor_7
sensor.philio_technology_corporation_psm021_slim_multisensor_luminance_7
sensor.philio_technology_corporation_psm021_slim_multisensor_motion_sensor_7
sensor.philio_technology_corporation_psm021_slim_multisensor_tamper_sensor_7
sensor.philio_technology_corporation_psm021_slim_multisensor_temperature_7

sensor.philio_technology_corporation_psm021_slim_multisensor_doorwindow_sensor_8
sensor.philio_technology_corporation_psm021_slim_multisensor_luminance_8
sensor.philio_technology_corporation_psm021_slim_multisensor_motion_sensor_8
sensor.philio_technology_corporation_psm021_slim_multisensor_tamper_sensor_8
sensor.philio_technology_corporation_psm021_slim_multisensor_temperature_8

And if there is several values of the same type in a node the instance id will be appended like this (three switches in the same device):
switch.fibaro_system_unknown_type0202_id1002_switch_9
switch.fibaro_system_unknown_type0202_id1002_switch_9_2
switch.fibaro_system_unknown_type0202_id1002_switch_9_3

stefan-jonasson commented:
This will still require all z-wave users to update their configurations. But the names will be intact.
Did not add `command_class` because it was no point, It could be several values with the same command_class like in my switch example.

Landrash commented:
As long as there's documentation with the update of the breaking changes there shouldn't be any bigger problems.  This is something I would have loved first time i setup my Z-wave devices since the naming didn't really help to differentiate between my devices =)

balloob commented:
I think it would make sense to create a Zwave class that contains entity_id, unique_id that we have all other classes inherit from. (python supports multiple inheritance)

Commit:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	homeassistant/components/light/zwave.py
#	homeassistant/components/switch/zwave.py

Commit:
Moved common z-wave node functionality to a common z-wave class

stefan-jonasson commented:
Moved all duplicated code to a common ancestor. Looks much better!

balloob commented on homeassistant/components/switch/zwave.py:
```diff
@@ -9,6 +9,9 @@
 from homeassistant.components.switch import SwitchDevice
+from homeassistant.components.zwave import ZWaveDeviceEntity
+
+DOMAIN = "switch"
```
You should import domain from the switch  component

Commit:
Updated Imports.
ZWave modules now import the domain from the base domain.

partofthething commented:
@stefan-jonasson,
> Did not add command_class because it was no point, It could be several values with the same command_class like in my switch example.

I think we need to have a command class number in the name after all. See my investigation into #1828. The `COMMAND_CLASS_METER` version of `sensor.aeotec_smart_energy_switch_power_4` is overwriting the `COMMAND_CLASS_SENSOR_MULTILEVEL` and preventing it from processing updates. In this case, it seems the device should send updates for both command classes (and maybe can be configured to do so), but anyway I think command classes need to be reflected in the name. 

Alternatively (and more seamlessly, since we don't have to break everyone's configs again), we could use an instance count and if there are more than one entities, we append a count number (>1). This looks like it is partially implemented in `homeassistant.components.zwave._object_id` but nothing seems to be incrementing the `value.instance` count. If we can just set that up, I'd be happy. Haven't found where to do it yet though.  The challenge here is that it seems that the different overlapping command classes can be discovered in different orders, so the COMMAND_CLASS-based name is more stable and robust. 

Perhaps a name like `sensor.aeotec_smart_energy_switch_power_50_4` for COMMAND_CLASS_METER (50) and  `sensor.aeotec_smart_energy_switch_power_49_4` for multi-level sensor. That's kind of ugly, but it's robust. Other ideas?

partofthething commented:
Yeah, using 
```
    object_id = "{}_{}_{}".format(slugify(_value_name(value)),
                               value.node.node_id, value.command_class)
```
totally makes things update better in #1828. But it's a very breaking change for all zwave stuff. And I don't really like the ugly command class numbers.
